File: sdk/flutter/dazzle_flutter/lib/src/agent/message.dart
// Copyright 2026 Ivan Aliaga <ivan.aliaga@urp.edu.pe>
// SPDX-License-Identifier: Apache-2.0
//
// Message + ChatTurn + ToolCall + Delta types. Same shape as
// `Message.kt` / `Message.swift` so JSON payloads round-trip cleanly
// between native and Flutter consumers of the same Dazzle store.

enum Role { system, user, assistant, tool }

extension RoleSerde on Role {
  String get wire => name; // lowercase matches Kotlin/Swift enum.
  static Role fromWire(String s) =>
      Role.values.firstWhere((r) => r.name == s, orElse: () => Role.user);
}

class ToolCall {
  final String id;
  final String name;
  final String arguments;   // raw JSON
  const ToolCall({required this.id, required this.name, required this.arguments});
}

/// Input to `LLMClient.complete` / `stream`. Mirrors the native
/// `Message` exactly so a Dazzle persistence file written by a Kotlin
/// app reads fine from Flutter.
class Message {
  final Role role;
  final String content;
  final List<ToolCall> toolCalls;
  final String? toolCallId;
  const Message({
    required this.role,
    required this.content,
    this.toolCalls = const [],
    this.toolCallId,
  });
}

/// A persisted turn in the chat agent's ContextStore. A superset of
/// `Message` with id + timestamp (so we can restore chronological order
/// on cold boot).
class ChatTurn {
  final String id;
  final Role role;
  final String text;
  final List<ToolCall> toolCalls;
  final String? toolCallId;
  final int timestamp; // millis since epoch

  ChatTurn({
    required this.id,
    required this.role,
    required this.text,
    this.toolCalls = const [],
    this.toolCallId,
    int? timestamp,
  }) : timestamp = timestamp ?? DateTime.now().millisecondsSinceEpoch;

  Message toMessage() => Message(
        role: role,
        content: text,
        toolCalls: toolCalls,
        toolCallId: toolCallId,
      );
}

/// Streaming events. Same four shapes the native SDK emits.
sealed class Delta {
  const Delta();
}

class DeltaText extends Delta {
  final String chunk;
  const DeltaText(this.chunk);
}

class DeltaToolCallStart extends Delta {
  final String id;
  final String name;
  const DeltaToolCallStart({required this.id, required this.name});
}

class DeltaToolCallArgs extends Delta {
  final String id;
  final String chunk;
  const DeltaToolCallArgs({required this.id, required this.chunk});
}

class DeltaEnd extends Delta {
  const DeltaEnd();
}

/// Final reply shape from `LLMClient.complete`.
sealed class Completion {
  const Completion();
}

class CompletionText extends Completion {
  final Message message;
  const CompletionText(this.message);
}

class CompletionToolCalls extends Completion {
  final Message message;
  const CompletionToolCalls(this.message);
}

/// Status a chat agent can be in — drives the UI's disabled-state.
enum AgentStatus { idle, thinking, streaming, toolCalling, error }

class StreamingMessage {
  final String text;
  final String? activeTool;
  const StreamingMessage({this.text = '', this.activeTool});
}

File: sdk/react-native/dazzle-react-native/src/agent/message.ts
// Copyright 2026 Ivan Aliaga <ivan.aliaga@urp.edu.pe>
// SPDX-License-Identifier: Apache-2.0

export type Role = 'system' | 'user' | 'assistant' | 'tool';

export interface ToolCall {
  readonly id: string;
  readonly name: string;
  readonly arguments: string;   // raw JSON
}

export interface Message {
  readonly role: Role;
  readonly content: string;
  readonly toolCalls?: ToolCall[];
  readonly toolCallId?: string;
}

export interface ChatTurn {
  readonly id: string;
  readonly role: Role;
  readonly text: string;
  readonly toolCalls?: ToolCall[];
  readonly toolCallId?: string;
  readonly timestamp: number;
}

export function turnToMessage(t: ChatTurn): Message {
  return {
    role: t.role,
    content: t.text,
    toolCalls: t.toolCalls,
    toolCallId: t.toolCallId,
  };
}

// Streaming deltas emitted by LLMClient.stream.
export type Delta =
  | { type: 'text'; chunk: string }
  | { type: 'toolCallStart'; id: string; name: string }
  | { type: 'toolCallArgs'; id: string; chunk: string }
  | { type: 'end' };

// Blocking completion shape.
export type Completion =
  | { type: 'text'; message: Message }
  | { type: 'toolCalls'; message: Message };

export type AgentStatus =
  | 'idle' | 'thinking' | 'streaming' | 'toolCalling' | 'error';

export interface StreamingMessage {
  readonly text: string;
  readonly activeTool: string | null;
}
